tree/30_generic_methods/directory_absent.cf

```diff
@@ -0,0 +1,85 @@
+#####################################################################################
+# Copyright 2017 Normation SAS
+#####################################################################################
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, Version 3.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+#####################################################################################
+
+# @name Directory absent
+# @description Ensure a directory's absence
+# @documentation If `recursive` is false, only an empty directory can be deleted.
+#
+# @parameter target     Directory to remove
+# @parameter recursive  Should deletion be recursive, "true" or "false" (defaults to "false")
+# @parameter_constraint recursive "allow_empty_string" : true
+#
+# @class_prefix directory_absent
+# @class_parameter target
+
+bundle agent directory_absent(target, recursive)
+{
+  vars:
+      "class_prefix" string => canonify("directory_absent_${target}");
+
+    target_exists::
+      # To check if the target is empty
+      "directory_content"    slist => lsdir("${target}/.", ".*", "false");
+      "directory_content_size" int => length("directory_content");
+
+  defaults:
+      "recursive"           string => "false", if_match_regex => "";
+
+  classes:
+      "target_exists"   expression => fileexists("${target}/.");
+      "is_recursive"    expression => strcmp("true", "${recursive}");
+    pass2::
+      "is_empty"        expression => strcmp("0", "${directory_content_size}");
+
+    any::
+      "pass3" expression => "pass2";
+      "pass2" expression => "pass1";
+      "pass1" expression => "any";
+
+  files:
+
+    pass2.target_exists.is_recursive::
+      "${target}/."
+        delete        => tidy,
+        file_select   => all,
+        depth_search  => recurse_with_base("inf"),
+        classes       => classes_generic("${class_prefix}_inside");
+
+    pass2.target_exists::
+      "${target}/."
+        delete        => tidy,
+        file_select   => all,
+        classes       => classes_generic("${class_prefix}"),
+        ifvarclass    => "!${class_prefix}_inside_error.(is_recursive|is_empty)";
+
+  methods:
+    pass3::
+      # If not empty, we need to trigger an error manually as the files promise will be kept
+      # Because of https://tracker.mender.io/browse/CFE-1570
+      "fail_if_not_empty"  usebundle  => _classes_failure("${class_prefix}"),
+                           ifvarclass => "target_exists.!is_recursive.!is_empty";
```

Comment:
Risk is the behaviour of the agent changes, and so we get a reports based on the expected "non logical" behaviour, rather than actual one.
Could you delete in the folder in pass1, and check in pass2 content of folder, to be future-proof?